apps/server/src/shared/domain/entity/permission-context.entity.ts

```diff
@@ -0,0 +1,106 @@
+				excludedPermissions: userDelta.excludedPermissions,
+			};
+		});
+	}
+
+	[userId: string]: {
+		includedPermissions: Permission[];
+		excludedPermissions: Permission[];
+	};
+}
+
+export interface IPermissionContextProperties {
+	name?: string | null;
+	userDelta?: UserDelta;
+	contextReference: ObjectId;
+	parentContext: PermissionContextEntity | null;
+}
+
+@Entity({ tableName: 'permission-context' })
+export class PermissionContextEntity extends BaseEntityWithTimestamps {
+	@Property()
+	@Unique()
+	contextReference: ObjectId;
+
+	@Property()
+	name: string | null;
+
+	@Index()
+	@Property()
+	userDelta: UserDelta;
+
+	@ManyToOne(() => PermissionContextEntity, { wrappedReference: true, fieldName: 'parentContext', nullable: true })
+	@Index()
+	_parentContext: Reference<PermissionContextEntity> | null;
+
+	constructor(props: IPermissionContextProperties) {
+		super();
+		this.contextReference = props.contextReference;
+		this._parentContext = props.parentContext ? new Reference(props.parentContext) : null;
+		this.name = props.name ?? null;
+		this.userDelta = props.userDelta ?? new UserDelta([]);
+	}
+
+	get parentContext(): Promise<PermissionContextEntity | null> {
+		if (!this._parentContext || this._parentContext.isInitialized()) {
+			return Promise.resolve(this._parentContext?.getEntity() ?? null);
+		}
+		return this._parentContext.load();
+	}
+
+	private resolveUserDelta(userId: User['id']): {
+		includedPermissions: Permission[];
+		excludedPermissions: Permission[];
+	} {
+		const userDelta = this.userDelta[userId] ?? { includedPermissions: [], excludedPermissions: [] };
+
+		return userDelta;
+	}
+
+	public async resolvedPermissions(userId: User['id']): Promise<Permission[]> {
+		const parent = await this.parentContext;
+		let parentPermissions: Permission[] = [];
+		if (parent) {
+			parentPermissions = await parent.resolvedPermissions(userId);
+		}
+		const userDelta = this.resolveUserDelta(userId);
+
+		const finalPermissions = parentPermissions
+			.concat(userDelta.includedPermissions)
+			.filter((permission) => !userDelta.excludedPermissions.includes(permission));
+
+		return [...new Set(finalPermissions)];
+	}
+
+	public async resolveFullPermissionMatrix(): Promise<Map<string, Permission[]>> {
+		const parent = await this.parentContext;
+		let parentPermissionMatrix: Map<string, Permission[]> = new Map();
+		if (parent) {
+			parentPermissionMatrix = await parent.resolveFullPermissionMatrix();
+		}
+
+		const permissionMatrix = new Map(parentPermissionMatrix);
+
+		Object.entries(this.userDelta).forEach(([userId, { includedPermissions, excludedPermissions }]) => {
+			const parentPermissions = parentPermissionMatrix.get(userId) ?? [];
+			const permissions = includedPermissions
+				.concat(parentPermissions)
+				.filter((permission) => !excludedPermissions.includes(permission));
+			permissionMatrix.set(userId, permissions);
+		});
+
+		return permissionMatrix;
+	}
+}
```

Comment:
where would the user role come into play?

We might have something like implicit perrmissions frrom the role, and then, explicit permission here?
Implicit
| actor (group / role / user) | entity (collection or document)  | action | allowed |
|--------|--------|--------|--------|
| student | column-node | read | true |
| student | column-node | write | false |
| teacher | column-node | write | true |
| userId | column-node-id | write | true |
| userId | card-node-id | read | false |